fdbserver/storageserver.actor.cpp

```diff
@@ -1200,6 +1200,67 @@ struct StorageServer : public IStorageMetricsService {
 
 	Optional<EncryptionAtRestMode> encryptionMode;
 
+	// Used for recording shard assignment history for auditStorage
+	std::vector<std::pair<Version, KeyRange>> assignedShardHistory;
```

Comment:
Can this handle unassigning ranges?

Reply:
Currently, assignedShardHistory records any shard change by changeServerKeys and changeServerKeysWithPhysicalShards which should be the only place to update the shard assignment to storage servers (including assigning ranges and unassigning ranges). However, the assignedShardHistory data structure itself currently does not distinguish the assigning ranges operations and unassigning ranges operations. This is sufficient for current usage: we want to compare local assignment and serverKey space at two different versions, and we only check there is any operation between the two versions. If yes, then give up comparing. In the future, we may want to replay the operation of shard assignment for the comparison. For this case, we will add operation type (i.e., shard assigning and shard unassigning) to assignedShardHistory. Let me change a name of assignedShardHistory to make it more clear.